### PR TITLE
add profile api

### PR DESCRIPTION
**Pull request by yaojiefeng**
Doc: https://wiki.cask.co/display/CE/CDAP+Provisioning

### CODE REVIEW

File: cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/ProfileScope.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.runtime.spi.profile;
+
+/**
+ * The scope of an artifact.
+ */
+public enum ProfileScope {
```
Comment (chtyim) [inline cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/ProfileScope.java]: You don't need this. Just use `EntityScope` class.

File: cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import co.cask.cdap.runtime.spi.provisioner.ProvisionerSpecification;
+
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Information of a profile. It encapsulates any information required to setup and teardown the program execution
+ * environment. A profile is identified by name and must be assigned a provisioner and its related configuration.
+ */
+public class Profile {
```
Comment (chtyim) [inline cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java]: does this belongs to SPI or app-fabric / proto ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                            @PathParam("namespace-id") String namespaceId,
+                            @PathParam("name") String profileName) throws NotFoundException {
+    // TODO: implement the method
+  }
+
+  /**
+   * List runs that the profile has been assigned to, either as the manual run profile or as a schedule profile.
+   */
+  @POST
+  @Path("/namespaces/{namespace-id}/profiles/{profile-name}/programs")
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java]: we're removing this endpoint and using metadata instead. I forgot to remove it from the wiki, but have removed it now.

File: cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java (hunk 2 above)
Comment (albertshau) [reply cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java]: I think this belongs in proto, as it's not used by anything in the spi.

File: cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  private final String name;
+  private final String description;
+  private final ProfileScope scope;
+  private final Map<String, String> primaryProperties;
```
Comment (albertshau) [inline cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java]: this and secondary shouldn't be here. It's just adding a 'required' boolean to the ProvisionerProperty class.

File: cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  private final Map<String, String> secondaryProperties;
+  private final ProvisionerSpecification provisionerSpecification;
```
Comment (albertshau) [inline cdap-runtime-spi/src/main/java/co/cask/cdap/runtime/spi/profile/Profile.java]: ProvisionerSpecification describes a provisioner. This is a different object that specifies the provisioner name and properties. It should match the response for the `/v3/namespaces/{namespace}/profiles/{profile}` endpoint.

File: cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProvisionerInfo.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Information about a provisioner name and properties.
+ */
+public class ProvisionerInfo {
+  private final String name;
+  private final String description;
```
Comment (albertshau) [inline cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProvisionerInfo.java]: this class doesn't need a description

File: cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProvisionerInfo.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  private final List<ProvisionerProperty> properties;
```
Comment (albertshau) [inline cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProvisionerInfo.java]: properties should just be a string string map. ProvisionerProperty is like a specification for each property. It is used in the provisioners rest apis, but not in the profile api. Properties here contains the values for each provisioner property.

**Review by albertshau — APPROVED** [review]
lgtm

**Pull request merged**